### ISSUE TITLE
'getsockname failed: Not a socket' when installing remote dataset via ssh on Windows 10

### ISSUE THREAD

**Issue by riedelmax** [datalad/datalad#5031]
#### What is the problem?
On Windows 10 `datalad -l debug install -s riedel@<ip-addres>:/home/riedel/data/rhizo` hangs with the last output being `getsockname failed: Not a socket`. Can't even kill the command with ctrl+c, only way is to close the PowerShell window.

The output is
```
[DEBUG] Command line args 1st pass. Parsed: Namespace() Unparsed: ['install', '-s', 'riedel@<ip-address>:/home/riedel/data/rhizo']
[DEBUG] Discovering plugins
[DEBUG] Building doc for <class 'datalad.local.subdatasets.Subdatasets'>
[DEBUG] Building doc for <class 'datalad.core.distributed.clone.Clone'>
[DEBUG] Building doc for <class 'datalad.distribution.get.Get'>
[DEBUG] Building doc for <class 'datalad.distribution.install.Install'>
[DEBUG] Parsing known args among ['C:\\Users\\M.Riedel\\Desktop\\wurzelsegmentierung\\datalad-test\\env\\Scripts\\datalad', '-l', 'debug', 'install', '-s', 'riedel@<ip-address>:/home/riedel/data/rhizo']
[DEBUG] Determined class of decorated function: <class 'datalad.distribution.install.Install'>
[DEBUG] Determined class of decorated function: <class 'datalad.core.distributed.clone.Clone'>
[DEBUG] Determined clone target path from source
[DEBUG] Resolved clone target path to: 'C:\Users\M.Riedel\Desktop\wurzelsegmentierung\datalad-test\data\rhizo'
Clone attempt:   0%|          | 0.00/1.00 [00:00<?, ? Candidate locations/s][DEBUG] Opening SSHConnection(ctrl_path=<<WindowsPath('C++71 chars++c8')>>, sshri=<<SSHRI(hostname++32 chars++el')>>) by calling ['ssh', '-fN', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=15m', '-o', 'ControlPath=C:\\Users\\M.Riedel\\AppData\\Local\\datalad.org\\datalad\\Cache\\sockets\\f70d05c8', 'riedel@<ip-address>']
getsockname failed: Not a socket
```

#### What steps will reproduce the problem?

- Create a datalad dataset on a remote machine `<ip-address>` at `<path>`
- `datalad -l debug install -s <username>@<ip-address>:<path>`

#### What version of DataLad are you using (run `datalad --version`)? On what operating system (consider running `datalad wtf`)?
I'm using datalad version 0.13.4 on Windows 10

```
# WTF
## configuration <SENSITIVE, report disabled by configuration>
## datalad
  - full_version: 0.13.4
  - version: 0.13.4
## dataset
  - id: 345ee3a3-7874-458c-90df-08b2269be0da
  - metadata: <SENSITIVE, report disabled by configuration>
  - path: C:\Users\M.Riedel\Desktop\wurzelsegmentierung\datalad-test
  - repo: AnnexRepo
## dependencies
  - annexremote: 1.4.3
  - appdirs: 1.4.4
  - boto: 2.49.0
  - cmd:7z: 16.04
  - cmd:annex: 8.20200815-g335aae266
  - cmd:bundled-git: UNKNOWN
  - cmd:git: 2.16.1.windows.4
  - cmd:system-git: 2.16.1.windows.4
  - cmd:system-ssh: for
  - humanize: 3.0.1
  - iso8601: 0.1.13
  - keyring: 21.4.0
  - keyrings.alt: 4.0.0
  - msgpack: 1.0.0
  - requests: 2.24.0
  - tqdm: 4.50.1
  - wrapt: 1.12.1
## environment
  - PATH: C:\Users\M.Riedel\Desktop\wurzelsegmentierung\datalad-test\env/Scripts;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\libnvvp;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\libnvvp;C:\ProgramData\DockerDesktop\version-bin;C:\Program Files\Docker\Docker\Resources\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\libnvvp;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v8.0\bin;C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v8.0\libnvvp;C:\Program Files\Microsoft MPI\Bin\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files\dotnet\;C:\Program Files\Git\cmd;C:\Program Files (x86)\GtkSharp\2.12\bin;C:\Program Files\PuTTY\;C:\Program Files (x86)\Graphviz2.38\bin;C:\Program Files\Inkscape;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Users\M.Riedel\cuda_cudnn7\bin;C:\Users\M.Riedel\opencv\build\x64\vc14\bin;C:\Program Files\cmake_dep\bin;C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin;C:\Program Files\make;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Microsoft VS Code\bin;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\ProgramData\chocolatey\bin;C:\Program Files\TortoiseSVN\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\NVIDIA Corporation\Nsight Compute 2019.5.0\;C:\Program Files\7-Zip;C:\Users\M.Riedel\Python\Python38\Scripts\;C:\Users\M.Riedel\Python\Python38\;C:\Users\M.Riedel\AppData\Local\Programs\Python\Launcher\;C:\Users\M.Riedel\AppData\Local\Microsoft\WindowsApps;C:\texlive\2017\bin\win32;C:\Program Files\Microsoft VS Code\bin;C:\tools\neovim\Neovim\bin;
## extensions
## git-annex
  - build flags:
    - Assistant
    - Webapp
    - Pairing
    - S3
    - WebDAV
    - TorrentParser
    - Feeds
    - Testsuite
  - dependency versions:
    - aws-0.21.1
    - bloomfilter-2.0.1.0
    - cryptonite-0.25
    - DAV-1.3.4
    - feed-1.2.0.1
    - ghc-8.6.5
    - http-client-0.5.14
    - persistent-sqlite-2.9.3
    - torrent-10000.1.1
    - uuid-1.3.13
    - yesod-1.6.0.1
  - key/value backends:
    - SHA256E
    - SHA256
    - SHA512E
    - SHA512
    - SHA224E
    - SHA224
    - SHA384E
    - SHA384
    - SHA3_256E
    - SHA3_256
    - SHA3_512E
    - SHA3_512
    - SHA3_224E
    - SHA3_224
    - SHA3_384E
    - SHA3_384
    - SKEIN256E
    - SKEIN256
    - SKEIN512E
    - SKEIN512
    - BLAKE2B256E
    - BLAKE2B256
    - BLAKE2B512E
    - BLAKE2B512
    - BLAKE2B160E
    - BLAKE2B160
    - BLAKE2B224E
    - BLAKE2B224
    - BLAKE2B384E
    - BLAKE2B384
    - BLAKE2BP512E
    - BLAKE2BP512
    - BLAKE2S256E
    - BLAKE2S256
    - BLAKE2S160E
    - BLAKE2S160
    - BLAKE2S224E
    - BLAKE2S224
    - BLAKE2SP256E
    - BLAKE2SP256
    - BLAKE2SP224E
    - BLAKE2SP224
    - SHA1E
    - SHA1
    - MD5E
    - MD5
    - WORM
    - URL
    - X*
  - local repository version: 8
  - operating system: mingw32 x86_64
  - remote types:
    - git
    - gcrypt
    - p2p
    - S3
    - bup
    - directory
    - rsync
    - web
    - bittorrent
    - webdav
    - adb
    - tahoe
    - glacier
    - ddar
    - git-lfs
    - hook
    - external
  - supported repository versions:
    - 8
  - upgrade supported from repository versions:
    - 2
    - 3
    - 4
    - 5
    - 6
    - 7
  - version: 8.20200815-g335aae266
## location
  - path: C:\Users\M.Riedel\Desktop\wurzelsegmentierung\datalad-test
  - type: dataset
## metadata_extractors
  - annex:
    - load_error: None
    - module: datalad.metadata.extractors.annex
    - version: None
  - audio:
    - load_error: No module named 'mutagen' [audio.py:<module>:17]
    - module: datalad.metadata.extractors.audio
  - datacite:
    - load_error: None
    - module: datalad.metadata.extractors.datacite
    - version: None
  - datalad_core:
    - load_error: None
    - module: datalad.metadata.extractors.datalad_core
    - version: None
  - datalad_rfc822:
    - load_error: None
    - module: datalad.metadata.extractors.datalad_rfc822
    - version: None
  - exif:
    - load_error: No module named 'exifread' [exif.py:<module>:16]
    - module: datalad.metadata.extractors.exif
  - frictionless_datapackage:
    - load_error: None
    - module: datalad.metadata.extractors.frictionless_datapackage
    - version: None
  - image:
    - load_error: No module named 'PIL' [image.py:<module>:16]
    - module: datalad.metadata.extractors.image
  - xmp:
    - load_error: No module named 'libxmp' [xmp.py:<module>:20]
    - module: datalad.metadata.extractors.xmp
## python
  - implementation: CPython
  - version: 3.8.5
## system
  - distribution:   10/10.0.19041/SP0/Multiprocessor Free
  - encoding:
    - default: utf-8
    - filesystem: utf-8
    - locale.prefered: cp1252
  - max_path_length: 259
  - name: Windows
  - release: 10
  - type: nt
  - version: 10.0.19041
```

**Comment by yarikoptic** [comment]
IMHO the problem starts from the initial establishing of the controlpath, you just don't see it due to `-fN`.  Filed an issue with conda-forge recipe for openssh, since it would be the most likely target ssh to be used: https://github.com/conda-forge/openssh-feedstock/issues/9

**Comment by riedelmax** [comment]
So is there another way I can install a dataset, that is accessible via ssh on Windows 10? Maybe deactivate the ssh multiplexing?

**Comment by mih** [comment]
I am working on a fix.

**Comment by mih** [comment]
I am hoping that https://github.com/datalad/datalad/pull/5042 introduces a working switch to turn of SSH multiplexing. This should make the underlying issues go away (by default on windows).... as long as you have key-based authentication set up and the respective key is passwordless or unlocked already (i.e. `ssh-agent` has it).

I have yet to verify on windows, but if you are interested, please give it a try.

**Comment by riedelmax** [comment]
Thank you very much for your effort. This fixed the original issue and I was able to install the datasat from the remote machine via ssh.

However, now I can't download the data with `datalad get .` and git-annex outputs some confusing warnings: `"You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time." ,"annex.sshcaching is not set to true"`

This looks like it has something to do with 
>as long as you have key-based authentication set up and the respective key is passwordless or unlocked already (i.e. `ssh-agent` has it).

I use key-based authentication, key and port are specified in my ssh config and I created a key without a password.

So right now I have no idea how to tackle this issue.

**Comment by yarikoptic** [comment]
try with `datalad get -J1 FILEs` ... if that doesn't work, try directly `git annex get -J1 FILEs`

Thanks for reporting on key-based auth gotcha -- I would have thought it would work, but I guess there is more to the story

**Comment by riedelmax** [comment]
Still not able to get the files..

The output of `git annex get -J1 ./Rhizo1_stats.txt` is
```
get Rhizo1_stats.txt (from origin...)
get Rhizo1_stats.txt (from origin...)
  You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time.

  annex.sshcaching is not set to true

  You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time.

  annex.sshcaching is not set to true

  rsync failed -- run git annex again to resume file transfer

  transfer failed

  Unable to access these remotes: origin

  Try making some of these repositories available:
        fd11db64-dd6b-473f-809b-664ed1934b25 -- riedel@GPU-Server:~/data/rhizo [origin]
failed
git-annex: get: 1 failed
```

I hope this helps

**Comment by yarikoptic** [comment]
hm... it is double weird that there are two
```
get Rhizo1_stats.txt (from origin...)
get Rhizo1_stats.txt (from origin...)
```
for the same file ( I do not think I saw such thing before)... and then all the "concurrency" messages.

what would be the output of `git config --list | grep annex` (i.e. any git config settings relating to annex) and would be so kind to rerun `git annex get -J1 --debug ./Rhizo1_stats.txt` ?

I also noted `- max_path_length: 259` . With `len('C:\Users\M.Riedel\Desktop\wurzelsegmentierung\datalad-test') = 58` taking 1/5th of the allowance, I am afraid that you might run into path length limitations sooner than later.

**Comment by mih** [comment]
Max path length 260 is standard: https://stackoverflow.com/questions/1880321/why-does-the-260-character-path-length-limit-exist-in-windows#:~:text=In%20the%20Windows%20API%20(with,and%20a%20terminating%20null%20character.

This was one of the main reasons to choose the MD5 backend by default.

Only rather recently this became an option to be relaxed on win10: https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#maximum-path-length-limitation

**Comment by riedelmax** [comment]
Output of `git config --list | select-string annex` (I'm on PowerShell):
```
remote.origin.annex-uuid=fd11db64-dd6b-473f-809b-664ed1934b25
annex.uuid=997fb402-c564-43fd-a686-d686d402b2dd
annex.sshcaching=false
annex.crippledfilesystem=true
annex.version=8
filter.annex.smudge=git-annex smudge -- %f
filter.annex.clean=git-annex smudge --clean -- %f
```

Output of `git annex get -J1 --debug ./Rhizo1_stats.txt`:
```
[2020-10-19 12:19:30.6888623] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","show-ref","git-annex"]
[2020-10-19 12:19:30.7074937] process done ExitSuccess
[2020-10-19 12:19:30.7123392] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","show-ref","--hash","refs/heads/git-annex"]
[2020-10-19 12:19:30.7490227] process done ExitSuccess
[2020-10-19 12:19:30.7570163] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","log","refs/heads/git-annex..c4ee8d7a4b29a06660c4fc0f1f8a2e2e72fa5bc3","--pretty=%H","-n1"]
[2020-10-19 12:19:30.7861374] process done ExitSuccess
[2020-10-19 12:19:30.7916927] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","hash-object","-w","--stdin-paths","--no-filters"]
[2020-10-19 12:19:30.795694] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch"]
[2020-10-19 12:19:30.8116957] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch-check=%(objectname) %(objecttype) %(objectsize)"]
[2020-10-19 12:19:30.8147168] feed: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","update-index","-z","--index-info"]
[2020-10-19 12:19:30.817694] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","diff-index","--raw","-z","-r","--no-renames","-l0","--cached","refs/heads/git-annex","--"]
[2020-10-19 12:19:30.8416944] process done ExitSuccess
[2020-10-19 12:19:30.8467299] process done ExitSuccess
[2020-10-19 12:19:30.8537213] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","symbolic-ref","-q","HEAD"]
[2020-10-19 12:19:30.8716924] process done ExitSuccess
[2020-10-19 12:19:30.8756946] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","show-ref","refs/heads/adjusted/master(unlocked)"]
[2020-10-19 12:19:30.8927088] process done ExitSuccess
[2020-10-19 12:19:30.9107244] read: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","ls-files","--stage","-z","--","./Rhizo1_stats.txt"]
[2020-10-19 12:19:30.9127253] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch-check=%(objectname) %(objecttype) %(objectsize)","--buffer"]
[2020-10-19 12:19:30.915692] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch=%(objectname) %(objecttype) %(objectsize)","--buffer"]
[2020-10-19 12:19:30.9196925] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch=%(objectname) %(objecttype) %(objectsize)","--buffer"]
get Rhizo1_stats.txt 
[2020-10-19 12:19:30.9516967] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch"]
[2020-10-19 12:19:30.9546932] process done ExitSuccess
[2020-10-19 12:19:30.9726938] chat: git ["--git-dir=.git","--work-tree=.","--literal-pathspecs","cat-file","--batch-check=%(objectname) %(objecttype) %(objectsize)"]
[2020-10-19 12:19:30.9736937] process done ExitSuccess
[2020-10-19 12:19:30.9766973] process done ExitSuccess
[2020-10-19 12:19:30.9776961] process done ExitSuccess
[2020-10-19 12:19:30.9776961] process done ExitSuccess
[2020-10-19 12:19:30.9786925] process done ExitSuccess
get Rhizo1_stats.txt (from origin...)
  You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time.

  annex.sshcaching is not set to true
[2020-10-19 12:19:31.6874109] process done ExitFailure 1
[2020-10-19 12:19:31.6884109] read: rsync ["--progress","--inplace","-e","'ssh' 'riedel@134.94.66.216' '-T' 'git-annex-shell ''sendkey'' ''/home/riedel/data/rhizo'' ''--debug'' ''MD5E-s227--33568811ff3c5572b0ab83c36b21a1f1.txt'' --uuid fd11db64-dd6b-473f-809b-664ed1934b25 ''--'' ''remoteuuid=997fb402-c564-43fd-a686-d686d402b2dd'' ''unlocked='' ''direct='' ''associatedfile=Rhizo1_stats.txt'' ''--'''","--","dummy:",".git/annex/tmp/MD5E-s227--33568811ff3c5572b0ab83c36b21a1f1.txt"]
[2020-10-19 12:19:31.753413] process done ExitSuccess
get Rhizo1_stats.txt (from origin...)
  You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time.

[2020-10-19 12:19:31.753413] process done ExitSuccess
  annex.sshcaching is not set to true

  You have enabled concurrency, but git-annex is not able to use ssh connection caching. This may result in multiple ssh processes prompting for passwords at the same time.

[2020-10-19 12:19:31.7554164] process done ExitSuccess
  annex.sshcaching is not set to true

  rsync failed -- run git annex again to resume file transfer

[2020-10-19 12:19:31.7564298] process done ExitSuccess
  transfer failed

  Unable to access these remotes: origin

  Try making some of these repositories available:
        fd11db64-dd6b-473f-809b-664ed1934b25 -- riedel@GPU-Server:~/data/rhizo [origin]
failed
[2020-10-19 12:19:31.7594204] process done ExitSuccess
git-annex: get: 1 failed
```


Seems like everything is done twice.

**Comment by yarikoptic** [comment]
Sorry for a somewhat lame followup but FWIW, there were a good number of fixes since 0.13.4 (also on git-annex side) which relate to windows and concurrency.  So if you @riedelmax could try out released datalad 0.13.7 and a recentish windows build from http://datasets.datalad.org/datalad/packages/windows/ -- would be appreciated since would guide us better on either we are yet to solve it ;)

**Comment by adswa** [comment]
If I'm not mistaken, #5031 should have fixed this. Please reopen if this is wrong :)
